File: src/util/Profiler.java
package util;

import interfaces.IProfiler;
import interfaces.ISearchAlgorithm;
import interfaces.ISortingAlgorithm;

public class Profiler implements IProfiler {
    @Override
    public long profileSearchTime(ISearchAlgorithm algorithm, int[] array, int element) {
        long start = System.nanoTime();
        algorithm.search(array, element);
        long end = System.nanoTime();
        return end - start;
    }

    @Override
    public long profileSortTime(ISortingAlgorithm algorithm, int[] array) {
        long start = System.nanoTime();
        algorithm.sort(array);
        long end = System.nanoTime();
        return end - start;
    }

    @Override
    public long profileMemoryUsed() {
        Runtime runtime = Runtime.getRuntime();
        runtime.gc();
        return runtime.totalMemory() - runtime.freeMemory();
    }
}

File: python/profiler.py
from iprofiler import IProfiler

class Profiler(IProfiler):
    def profile_search_time(self, algorithm, array, element):
        import time
        start = time.perf_counter_ns()
        algorithm.search(array, element)
        end = time.perf_counter_ns()
        return end - start

    def profile_sort_time(self, algorithm, array):
        import time
        start = time.perf_counter_ns()
        algorithm.sort(array)
        end = time.perf_counter_ns()
        return end - start

    def profile_memory_used(self):
        import os, psutil
        process = psutil.Process(os.getpid())
        return process.memory_info().rss